src/terms/Hamiltonian.jl

```diff
@@ -70,15 +66,22 @@ end
 end
 Base.:*(H::HamiltonianBlock, ψ) = mul!(similar(ψ), H, ψ)
 
+# Returns a fast path hamiltonian if eligible, nothing if not
+function fast_hblock(H::HamiltonianBlock)
+    length(H.optimized_operators) == 3 || return nothing
+    fourier_ops = filter(o -> o isa FourierMultiplication, H.optimized_operators)
+    real_ops = filter(o -> o isa RealSpaceMultiplication, H.optimized_operators)
+    nonlocal_ops = filter(o -> o isa NonlocalOperator, H.optimized_operators)
+    (length(fourier_ops) == length(real_ops) == length(nonlocal_ops) == 1) || return nothing
+    (fourier_op=only(fourier_ops), real_op=only(real_ops), nonlocal_op=only(nonlocal_ops), H=H)
+end
+
 # Fast version, specialized on DFT models with just 3 operators: real, fourier and nonlocal
 # Minimizes the number of allocations
-@views function mul!_fast(Hψ::AbstractArray,
-                          H::HamiltonianBlock,
-                          ψ::AbstractArray)
-    fourier_op = only(filter(o -> o isa FourierMultiplication, H.optimized_operators))
-    real_op = only(filter(o -> o isa RealSpaceMultiplication, H.optimized_operators))
-    nonlocal_op = only(filter(o -> o isa NonlocalOperator, H.optimized_operators))
-
+@views function fast_hblock_mul!(Hψ::AbstractArray,
+                                 fast_hblock,
```

Comment:
> Just add mul!(H\psi, H, \spi) = hblock_mul!(H\psi, H, fast_hblock(H), \psi) and than just specialise this function for the case where the second arg is nothing or when it is a named tuple.

I'd rather not do that, because that enshrines the hack... Right now it's clear that the official function is mul!, and that it internally does some hacks which are implementation details. Whereas your version hides that behind one more level of dispatch. Just because julia has dispatch doesn't mean we have to use it :p We can add underscores if you want to make that more explicit?

Reply:
Nah all right. I'm not happy with this, since it feels an awful lot like we are reimplementing a language feature. As I said to me such kind of specialisation to give special code paths for special cases is exactly what multiple dispatch is for. Sure you don't have to use it, but structure and patterns aid thinking ... could be there is more to be gained elsewhere. But whatever ... it's a hack first of all and probably not worth discussing endlessly.

Reply:
Structure is good for organizing large things. For implementation of small things, plain old C/fortran-style is better I think.